children/schema.py

```diff
@@ -224,17 +224,30 @@ def resolve_active_internal_and_ticket_system_enrolments(self, info, **kwargs):
         active_occurrences = Occurrence.objects.upcoming_with_ongoing()
         internal_enrolments = self.enrolments.filter(
             occurrence__in=active_occurrences
-        ).annotate(time=F("occurrence__time"))
-        ticket_system_passwords = self.ticket_system_passwords.filter(
-            event__occurrences__in=active_occurrences
-        ).annotate(time=Max("event__occurrences__time"))
+        ).annotate(
+            # Technically to be 100% correct we should use the occurrence's end time
+            # instead of start time for sorting because for Ticketmaster enrolments the
+            # event's end time is used, but doing that would be way more complicated
+            # and the difference not matter at all in practice.
+            time=F("occurrence__time"),
+            published_at=F("occurrence__event__published_at"),
+        )
+        ticket_system_passwords = (
+            self.ticket_system_passwords.event_upcoming_or_ongoing().annotate(
+                time=F("event__ticket_system_end_time"),
+                published_at=F("event__published_at"),
+            )
+        )
 
+        datetime_max = make_aware(datetime(9999, 2, 18))
         return sorted(
             (
                 *EnrolmentNode.get_queryset(internal_enrolments, info),
                 *TicketmasterEnrolmentNode.get_queryset(ticket_system_passwords, info),
             ),
-            key=lambda e: e.time,
+            # Sort events without ticket system end time as last ones and then sort
+            # them by published at.
+            key=lambda e: (e.time or datetime_max, e.published_at),
```

Comment:
There's [datetime.max](https://docs.python.org/3/library/datetime.html#datetime.datetime.max) which gives `datetime(MAXYEAR, 12, 31, 23, 59, 59, 999999, tzinfo=None)` but then that might break when changing the timezone as it might go over the maximum. Hmm...

One possibility to remove the datetime_max altogether if it e.time sorting is enough for those that have it:
```python
key=lambda e: (e.time is None, e.time or e.published_at),
```
but if they should also be sorted secondarily by e.published then:
```python
key=lambda e: (e.time is None, e.time or e.published_at, e.published_at),
```
but that gets complicated.

What's the reason for using 9999-02-18? Why not 9999-12-31?

Reply:
Yeah, `datetime.max` did not work, and neither did 31.12. (date value out of range). 18.2. is just an arbitrary choice, and if that happens to be someone's birthday, it is just a coincidence.

Reply:
Hmm... trying it out in django shell:
```python
make_aware(datetime(9999, 12, 31))

/usr/local/lib/python3.9/site-packages/django/utils/timezone.py in make_aware(value, timezone, is_dst)
    240     if _is_pytz_zone(timezone):
    241         # This method is available for pytz time zones.
--> 242         return timezone.localize(value, is_dst=is_dst)
    243     else:
    244         # Check that we won't overwrite the timezone of an aware datetime.

/usr/local/lib/python3.9/site-packages/pytz/tzinfo.py in localize(self, dt, is_dst)
    321         possible_loc_dt = set()
    322         for delta in [timedelta(days=-1), timedelta(days=1)]:
--> 323             loc_dt = dt + delta
    324             idx = max(0, bisect_right(
    325                 self._utc_transition_times, loc_dt) - 1)

OverflowError: date value out of range
```
So pytz's [localize](https://github.com/stub42/pytz/blob/df59d3af429d46ee575ab5d3a7fe3f8cd49a74bc/src/pytz/tzinfo.py#L322) which is called by django's make_aware adds one day to the value which goes over datetime.max. Ok, now I understand.

Reply:
Changed it to `make_aware(datetime.max, timezone=timezone.utc)`.